File: parser/src/main/java/io/ticofab/androidgpxparser/parser/domain/Track.java
package io.ticofab.androidgpxparser.parser.domain;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Track {
    private final String mTrackName;
    private final List<TrackSegment> mTrackSegments;
    private final String mTrackDesc;
    private final String mTrackCmt;
    private final String mTrackSrc;
    private final Integer mTrackNumber;
    private final Link mTrackLink;
    private final String mTrackType;

    private Track(Builder builder) {
        mTrackName = builder.mTrackName;
        mTrackDesc = builder.mTrackDesc;
        mTrackCmt = builder.mTrackCmt;
        mTrackSrc = builder.mTrackSrc;
        mTrackNumber = builder.mTrackNumber;
        mTrackSegments = Collections.unmodifiableList(new ArrayList<>(builder.mTrackSegments));
        mTrackLink = builder.mTrackLink;
        mTrackType = builder.mTrackType;
    }

    public String getTrackName() {
        return mTrackName;
    }

    public String getTrackDesc() {
        return mTrackDesc;
    }

    public String getTrackCmt() {
        return mTrackCmt;
    }

    public String getTrackSrc() {
        return mTrackSrc;
    }

    public Integer getTrackNumber() {
        return mTrackNumber;
    }

    public Link getTrackLink() {
        return mTrackLink;
    }

    public String getTrackType() {
        return mTrackType;
    }

    public List<TrackSegment> getTrackSegments() {
        return mTrackSegments;
    }

    public static class Builder {
        private String mTrackName;
        private List<TrackSegment> mTrackSegments;
        private String mTrackDesc;
        private String mTrackCmt;
        private String mTrackSrc;
        private Integer mTrackNumber;
        private Link mTrackLink;
        private String mTrackType;

        public Builder setTrackName(String trackName) {
            mTrackName = trackName;
            return this;
        }

        public Builder setTrackDesc(String trackDesc) {
            mTrackDesc = trackDesc;
            return this;
        }

        public Builder setTrackSegments(List<TrackSegment> trackSegments) {
            mTrackSegments = trackSegments;
            return this;
        }

        public Builder setTrackCmt(String trackCmt) {
            mTrackCmt = trackCmt;
            return this;
        }

        public Builder setTrackSrc(String trackSrc) {
            mTrackSrc = trackSrc;
            return this;
        }

        public Builder setTrackNumber(Integer trackNumber) {
            mTrackNumber = trackNumber;
            return this;
        }

        public Builder setTrackLink(Link link) {
            mTrackLink = link;
            return this;
        }

        public Builder setTrackType(String type) {
            mTrackType = type;
            return this;
        }

        public Track build() {
            return new Track(this);
        }


    }
}

File: parser/src/main/kotlin/io/ticofab/androidgpxparser/parser/domain/Track.kt
package io.ticofab.androidgpxparser.parser.domain

class Track private constructor(builder: Builder) {
    val trackName = builder.trackName
    val trackSegments = builder.trackSegments?.toList()
    val trackDesc = builder.trackDesc
    val trackCmt = builder.trackCmt
    val trackSrc = builder.trackSrc
    val trackNumber = builder.trackNumber
    val trackLink = builder.trackLink
    val trackType = builder.trackType

    class Builder {
        var trackName: String? = null
            private set
        var trackSegments: List<TrackSegment>? = null
            private set
        var trackDesc: String? = null
            private set
        var trackCmt: String? = null
            private set
        var trackSrc: String? = null
            private set
        var trackNumber: Int? = null
            private set
        var trackLink: Link? = null
            private set
        var trackType: String? = null
            private set

        fun setTrackName(trackName: String?) = apply { this.trackName = trackName }

        fun setTrackDesc(trackDesc: String?) = apply { this.trackDesc = trackDesc }

        fun setTrackSegments(trackSegments: List<TrackSegment>?) = apply { this.trackSegments = trackSegments }

        fun setTrackCmt(trackCmt: String?) = apply { this.trackCmt = trackCmt }

        fun setTrackSrc(trackSrc: String?) = apply { this.trackSrc = trackSrc }

        fun setTrackNumber(trackNumber: Int?) = apply { this.trackNumber = trackNumber }

        fun setTrackLink(link: Link?) = apply { trackLink = link }

        fun setTrackType(type: String?) = apply { trackType = type }

        fun build() = Track(this)
    }
}